staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go

```diff
@@ -145,6 +153,29 @@ func (f *FakeClock) After(d time.Duration) <-chan time.Time {
 	return ch
 }
 
+// AfterFunc is the Fake version of time.AfterFunc(d, callback).
+func (f *FakeClock) AfterFunc(d time.Duration, cb func()) Timer {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	stopTime := f.time.Add(d)
+	ch := make(chan time.Time, 1) // Don't block!
+
+	go func() {
+		<-ch
+		cb()
+	}()
```

Comment:
no goroutines in fake clocks... they mean that after calling time.Step(), our callback will be called asynchronously, which defeats the deterministic aspect of a fake clock. suggest something like the following instead:

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go b/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go
index 5346b66421a..0e42c9c00b9 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go
@@ -103,6 +103,7 @@ type fakeClockWaiter struct {
 	stepInterval  time.Duration
 	skipIfBlocked bool
 	destChan      chan time.Time
+	afterFunc     func()
 }
 
 // NewFakePassiveClock returns a new FakePassiveClock.
@@ -160,16 +161,12 @@ func (f *FakeClock) AfterFunc(d time.Duration, cb func()) Timer {
 	stopTime := f.time.Add(d)
 	ch := make(chan time.Time, 1) // Don't block!
 
 	timer := &fakeTimer{
 		fakeClock: f,
 		waiter: fakeClockWaiter{
 			targetTime: stopTime,
 			destChan:   ch,
+			afterFunc:  cb,
 		},
 	}
 	f.waiters = append(f.waiters, timer.waiter)
@@ -241,6 +238,9 @@ func (f *FakeClock) setTimeLocked(t time.Time) {
 			} else {
 				w.destChan <- t
 			}
+			if w.afterFunc != nil {
+				w.afterFunc()
+			}
 
 			if w.stepInterval > 0 {
 				for !w.targetTime.After(t) {
```

Reply:
makes sense. Applied suggested fix.